timemachine/md/hrex.py

```diff
@@ -69,16 +157,75 @@ def attempt_neighbor_swaps(
         log_q: Callable[[ReplicaIdx, StateIdx], float],
         n_swap_attempts: int,
     ) -> Tuple["HREX[Replica]", List[Tuple[int, int]]]:
-        move = BatchedMixtureOfMoves(
-            n_swap_attempts, [NeighborSwapMove(log_q, s_a, s_b) for s_a, s_b in neighbor_pairs]
-        )
+        """Run a batch of swap attempts.
+
+        See :py:meth:`attempt_neighbor_swaps_fast` for a more efficient implementation with a similar signature.
+        Note that these methods do not generate compatible random sequences.
+
+        Parameters
+        ----------
+        neighbor_pairs : Sequence[Tuple[StateIdx, StateIdx]]
+            pairs of states between which to attempt swaps
+
+        log_q : Callable[[ReplicaIdx, StateIdx], float]
+            function to compute the unnormalized probability of a given replica in a given state
+
+        n_swap_attempts : int
+            number of individual swap attempts, each between a randomly-selected neighbor pair
 
-        replica_idx_by_state = move.move(list(self.replica_idx_by_state))
+        Returns
+        -------
+        Tuple[HREX[Replica], List[Tuple[int, int]]]
+            Updated HREX state, list of (accepted, proposed) counts by neighbor pair
+        """
+        move = MixtureOfMoves([NeighborSwapMove(log_q, s_a, s_b) for s_a, s_b in neighbor_pairs])
+
+        replica_idx_by_state = move.move_n(list(self.replica_idx_by_state), n_swap_attempts)
 
         fraction_accepted_by_pair = list(zip(move.n_accepted_by_move, move.n_proposed_by_move))
 
         return HREX(self.replicas, replica_idx_by_state), fraction_accepted_by_pair
 
+    def attempt_neighbor_swaps_fast(
+        self,
+        neighbor_pairs: Sequence[Tuple[StateIdx, StateIdx]],
+        log_q_kl: ArrayLike,
+        n_swap_attempts: int,
+    ) -> Tuple["HREX[Replica]", List[Tuple[int, int]]]:
+        """Run a batch of swap attempts.
+
+        See :py:meth:`attempt_neighbor_swaps` for a (typically slower) reference version.
+        Note that these methods do not generate compatible random sequences.
+
+        Parameters
+        ----------
+        neighbor_pairs : Sequence[Tuple[StateIdx, StateIdx]]
+            pairs of states between which to attempt swaps
+
+        log_q_kl : ArrayLike
+            (n_replicas, n_states) array with the (r, s) element giving the unnormalized probability of replica r in state s
+
+        n_swap_attempts : int
```

Comment:
Tangent: Not in diff, but I just noticed a small inconsistency: `n_swap_attempts` is `K^4` in docstring, `K^3` in implementation https://github.com/proteneer/timemachine/blob/aaa5868db4ef8f2c7c789aa207bb1bbaef003206/timemachine/md/hrex.py#L367-L384

Reply:
Good catch. Fixed up docstring in https://github.com/proteneer/timemachine/pull/1294/commits/8f5e4492cb91794d5fca49216597c34ba596cf96.